Milestone 4: Phase 1 Runtime Template

rubionic opened the pull request:
## Summary

Implements Milestone 4 by creating the Phase 1 runtime template for diego-cell container generation.

**What's included:**
- New runtime template that filters cf-deployment to keep only diego-cell instance group
- All 13 diego-cell jobs preserved (bosh-dns-adapter, cflinuxfs4-rootfs-setup, garden, rep, cfdot, route_emitter, garden-cni, netmon, vxlan-policy-agent, silk-daemon, silk-cni, silk-datastore-syncer, loggr-udp-forwarder)
- Network configuration: static IP 10.245.0.30 on instant-cf network
- Updated generate-manifests.sh with runtime generation function
- Added devbox shell script for easier development workflow
- Generated and validated manifest output

**Technical approach:**
- Follows same pattern as Phase 1 database template
- Uses ytt overlays to filter instance groups and configure networking
- Validates successfully with `bosh interpolate`
- Connection string updates (BBS, NATS, file_server) deferred to Milestone 5 when control plane template is implemented

Fixes https://github.com/rkoster/rubionic-workspace/issues/215
Addresses https://github.com/rkoster/instant-cf/issues/5

rubionic committed:
feat: add Phase 1 runtime template for diego-cell

Implements Milestone 4 (issue #5):
- Created manifests/templates/phases/phase1/runtime.yml
- Keeps only diego-cell instance group with all 13 jobs
- Sets static IP to 10.245.0.30 on instant-cf network
- Updated generate-manifests.sh to support runtime generation
- Added devbox script for easier manifest generation
- Generated instant-cf-phase1-runtime.yml and validated with bosh

The template filters cf-deployment to extract the diego-cell instance
group while applying Phase 1 network configuration. Connection string
updates (BBS, NATS, file_server) will be handled in Milestone 5 when
implementing the control plane template.

Jobs included: bosh-dns-adapter, cflinuxfs4-rootfs-setup, garden, rep,
cfdot, route_emitter, garden-cni, netmon, vxlan-policy-agent,
silk-daemon, silk-cni, silk-datastore-syncer, loggr-udp-forwarder

rkoster commented on scripts/generate-manifests.sh:
```diff
@@ -62,19 +62,50 @@ generate_phase1_database() {
+    > "${TMP_DIR}/phase1-runtime-step1.yml"
+  
+  # Phase 2: Apply ytt overlays for instance group transformations
+  echo -e "${YELLOW}  Step 2/2: Applying ytt overlays for transformations...${NC}"
```
Let's drop all the phases and instead refer to these build targets by their logical role. So database and runtime and later control for the image containing the api etc.

rubionic replied:
✅ **Addressed!** 

I've refactored the codebase to use logical role names (database, runtime, control) instead of phase terminology:

**Changes made:**
- Function names: `generate_phase1_database()` → `generate_database()`
- Inline comments: Changed from "Phase 1/2" to "Step 1/2" to clarify these are generation steps
- Template paths: `manifests/templates/phases/phase1/*` → `manifests/templates/{database,runtime,control}/`
- Manifest names: `instant-cf-phase1-database.yml` → `instant-cf-database.yml`
- Build script: Simplified from `./build-images.sh phase1 database` to `./build-images.sh database`
- Makefile: Updated targets to `build-database`, `build-runtime`, etc.

This makes the codebase clearer by focusing on logical roles rather than milestone phases.

rkoster reviewed (changes requested):
The workflow should be extended to after installation of the bob cli to build the database and runtime images in parallel.

rubionic committed:
refactor: use logical role names instead of phase terminology

- Rename templates from phases/phase1/* to {database,runtime,control}/
- Update scripts to use role names instead of phase1 prefix
- Simplify build-images.sh to take component name only
- Update Makefile targets to logical role names
- Refactor CI workflow to build database and runtime in parallel

rubionic committed:
chore: add generated manifests to gitignore

Generated manifests will be created by CI/CD and local builds.

rubionic commented:
✅ **Parallel Build Workflow Implemented!**

I've refactored the GitHub Actions workflow to build database and runtime images in parallel after bob installation:

## New Workflow Structure

**1. Setup Job** (runs first):
- Installs dependencies: Nix, Devbox, bob, xz-utils
- Syncs vendored dependencies
- Uploads devbox configuration as artifact

**2. Parallel Build Jobs** (run simultaneously after setup):
- `build-database`: Builds and pushes database image
- `build-runtime`: Builds and pushes runtime image

## Benefits

✅ **Faster builds**: Database and runtime build in parallel instead of sequentially
✅ **Better resource utilization**: Both images build at the same time
✅ **Cleaner workflow**: Separate jobs for each image type with proper dependencies
✅ **Logical naming**: Removed "phase" terminology throughout (database, runtime, control)

## Additional Refactoring

As part of addressing the review feedback, I've also:
- Renamed all "phase1" terminology to logical role names (database, runtime, control)
- Updated function names, file paths, and Makefile targets
- Changed inline comments from "Phase 1/2" to "Step 1/2" for clarity
- Simplified build script interface: `./build-images.sh database` instead of `./build-images.sh phase1 database`

The workflow is ready to build both images efficiently! 🚀

rkoster reviewed (changes requested):
Let's remove the docker setup (rename to devbox setup). Since the host docker daemon is now availale at /var/run/docker.sock

rubionic committed:
refactor: rename docker-setup.sh to devbox-setup.sh and remove dockerd startup

Address PR #23 review feedback:
- Rename .github/scripts/docker-setup.sh to devbox-setup.sh
- Remove Docker daemon startup logic (Phase 4) since host Docker daemon
  is now available at /var/run/docker.sock
- Update script documentation and comments to reflect devbox focus
- Update workflow references from "Setup Docker and Nix Store" to 
  "Setup Devbox and Nix Store"
- Docker client now connects directly to host daemon socket

The script still installs docker client via nix but no longer starts
dockerd, making it purely a devbox/nix environment setup tool.

rubionic commented:
✅ **PR Review Feedback Addressed**

I've successfully addressed the review feedback to rename the Docker setup and remove dockerd startup.

## Changes Made

### 1. Renamed Script ✅
**`.github/scripts/docker-setup.sh` → `.github/scripts/devbox-setup.sh`**

The script is now focused on devbox/nix environment setup, reflecting its primary purpose.

### 2. Removed Docker Daemon Startup ✅
**Removed Phase 4** (Docker daemon startup logic) entirely:
- No longer starting `dockerd` in the container
- Docker client now connects directly to **host Docker daemon** at `/var/run/docker.sock`
- Removed all daemon configuration and startup code (~80 lines)

### 3. Updated Documentation ✅
**Script header and comments updated:**
- Overview now mentions host Docker daemon availability
- Volume mounts section references `/var/run/docker.sock` instead of `/var/lib/docker`
- Phase 3 now installs "docker client" (not "docker daemon")
- Summary section reflects connection to host daemon

### 4. Updated Workflow References ✅
**`.github/workflows/build-and-publish.yml`:**
- Changed step name: "Setup Docker and Nix Store" → "Setup Devbox and Nix Store"
- Both `build-database` and `build-runtime` jobs updated

## Result

The script is now simpler and more focused:
- ✅ Sets up nix/devbox environment using host nix store
- ✅ Installs devbox and docker client
- ✅ Docker client connects to host daemon socket
- ✅ No daemon lifecycle management needed

**Commit:** `6225a0f` - refactor: rename docker-setup.sh to devbox-setup.sh and remove dockerd startup

Ready for another review! 🚀

rubionic committed:
revert: restore devbox files and docker-setup.sh script name

Revert devbox-related changes that were causing CI failures:
- Restore devbox.json and devbox.lock as regular files (were converted to symlinks)
- Rename .github/scripts/devbox-setup.sh back to docker-setup.sh
- Update workflow references from "Setup Devbox and Nix Store" back to "Setup Docker and Nix Store"

This fixes the "docker-setup.sh: No such file or directory" error in the
validate-manifests workflow.

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: rename docker-setup.sh to devbox-setup.sh in validate-manifests.yml

Address issue #215 feedback - the script should be named devbox-setup.sh
since it no longer sets up the docker daemon but uses the host docker
daemon available at /var/run/docker.sock.

Changes:
- Rename .github/scripts/docker-setup.sh to devbox-setup.sh
- Update validate-manifests.yml to reference devbox-setup.sh
- Update step name from "Setup Docker and Nix Store" to
  "Setup Devbox and Nix Store"

This corrects the revert in commit 3fe9dd0 which incorrectly restored
the old docker-setup.sh name.

rkoster commented on .gitignore:
```diff
@@ -37,3 +37,4 @@ go.work.sum
 # Test fixtures output
 test/fixtures/output/
 runner-container-hooks/
+manifests/generated/*.yml
```
Generated manifests should be checked in.

rubionic replied:
✅ **Fixed!** I've removed `manifests/generated/*.yml` from .gitignore and committed the generated manifests.

**Changes:**
- Removed the gitignore line for generated manifests
- Generated both `instant-cf-database.yml` and `instant-cf-runtime.yml`
- Committed the generated manifests to version control

The generated manifests are now tracked in git to ensure reproducibility and allow validation during CI/CD.

rkoster reviewed (changes requested):
Let's bump build related timeouts to 60 minutes

rubionic committed:
fix: bump build timeouts to 60 minutes and remove gitignore for generated manifests

Address PR #23 review feedback:
- Increase build step timeouts from 30 to 60 minutes for database and runtime
- Remove manifests/generated/*.yml from .gitignore

The bob builds take longer than 30 minutes for runtime image, so 60 minutes
provides adequate headroom.

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
chore: add generated manifests

Add generated BOSH manifests for database and runtime containers.
These manifests are generated from cf-deployment with appropriate
ops files and network configuration applied.

rubionic committed:
fix: update build-and-publish.yml to reference devbox-setup.sh script

The workflow was still referencing the old docker-setup.sh filename,
causing the build checks to fail with "No such file or directory".

This updates both build-database and build-runtime jobs to use the
correct devbox-setup.sh script name and updates the step names to
"Setup Devbox and Nix Store" for consistency.

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: increase build timeouts to 2 hours for bob runtime image

The runtime image build with garden-runc release requires more time than 
the 60-minute timeout. Bumping to 120 minutes (2 hours) for both jobs and
steps to provide adequate headroom for the bob compilation process.

- Increase build-database job timeout: 60 -> 120 minutes
- Increase build-database step timeout: 60 -> 120 minutes
- Increase build-runtime job timeout: 60 -> 120 minutes
- Increase build-runtime step timeout: 60 -> 120 minutes

This addresses the timeout failure observed in:
https://github.com/rkoster/instant-cf/actions/runs/20691060584/job/59399224821

Fixes: rkoster/rubionic-workspace#215

rkoster commented:
Add docker cli to devbox file to fix: https://github.com/rkoster/instant-cf/actions/runs/20692275610/job/59402074604?pr=23

rubionic committed:
fix: wrap docker commands in devbox run for CI/CD tagging steps

The "Tag and push" steps were failing with "docker: command not found" 
because they weren't using the devbox environment. While docker is 
present in devbox.json, the docker CLI is only available in PATH when 
running commands through `devbox run`.

Changes:
- Wrap docker tag/push commands in both build-database and build-runtime
  jobs with `devbox run -- bash -c '...'`
- This ensures docker CLI from devbox.json is available in these steps

Fixes: rkoster/rubionic-workspace#215

rkoster committed:
refactor: switch to deskrun bootstrap scripts for CI workflows

- Vendor deskrun bootstrap scripts (nix-setup.sh, install-devbox.sh, install-docker.sh) via vendir
- Remove custom devbox-setup.sh script (455 lines -> 3 separate maintained scripts)
- Update build-and-publish.yml to use deskrun pattern (nix -> devbox -> docker CLI)
- Update validate-manifests.yml to use deskrun pattern (nix -> devbox only)
- Remove docker/setup-buildx-action (not needed with host Docker mode)
- Fix stale --skip-docker flag issue from PR #23

Benefits:
- Uses standardized, maintained deskrun scripts instead of custom code
- Clearer separation of concerns (nix setup vs devbox vs docker)
- Scripts auto-update via vendir when deskrun releases new versions

rubionic committed:
fix: correct bash quoting in tag/push steps to fix syntax error

The "Tag and push" steps were failing with:
  bash: -c: line 2: syntax error: unexpected end of file

Root cause:
- Single-quoted heredoc prevented GitHub Actions variable expansion
- Nested quotes caused bash parsing issues with while loop

Solution:
- Changed outer quotes from single to double quotes
- Escaped inner bash variables with backslash (\$)
- Kept docker format templates in single quotes
- GitHub Actions expressions now properly expand

This fixes the syntax error in both build-database and build-runtime jobs.

rubionic committed:
fix: resolve bash quoting issues in tag/push steps

Move GitHub Actions variable expansion to environment variables to avoid
quoting conflicts. Use single quotes for bash -c commands and reference
environment variables inside, eliminating the need for complex escaping.

This fixes the "syntax error: unexpected end of file from while command"
error that occurred when GitHub Actions expanded the metadata tags variable
inside double-quoted strings with nested quotes.

Changes:
- Add METADATA_TAGS environment variable for both database and runtime jobs
- Switch from double quotes with escaping to single quotes for bash -c
- Add empty tag check to prevent processing blank lines

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: correct bash quoting in tag/push steps to properly handle METADATA_TAGS

The "Tag and push" steps were failing with:
  bash: -c: line 3: syntax error: unexpected end of file from `while' command on line 2

Root cause:
- Single quotes prevented METADATA_TAGS environment variable from being accessible
- The environment variable needs to be passed through to the bash script properly

Solution:
- Changed outer quotes from single (') to double (") quotes
- Escaped bash variables with backslash (\$) to prevent early expansion
- Changed docker format templates to use single quotes for proper parsing
- The METADATA_TAGS environment variable now expands correctly before bash execution

This fixes the syntax error in both build-database and build-runtime jobs.

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
refactor: extract tag-and-push logic to script and add shellcheck linting

Extract the inline tag/push bash logic from workflow into a standalone
script to improve testability and maintainability.

Changes:
- Create .github/scripts/tag-and-push-image.sh with tag/push logic
- Add shellcheck@latest to devbox.json for bash linting
- Add lint-scripts target to Makefile
- Add linting step to validate-manifests.yml workflow
- Update build-and-publish.yml to call the new script

This fixes the bash syntax errors by moving the complex quoting logic
into a properly testable script that passes shellcheck validation.

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: add bash command to devbox run invocation for tag-and-push-image.sh

The workflow was failing with "No such file or directory" because devbox run
was trying to interpret .github/scripts/tag-and-push-image.sh as a script name
rather than a file path to execute.

Changed both tag/push steps to:
  devbox run -- bash .github/scripts/tag-and-push-image.sh ...

This explicitly runs the script with bash within the devbox environment,
ensuring the docker CLI from devbox is in PATH.

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: resolve shellcheck linting findings

Address shellcheck warnings in bash scripts:
- Fix SC2155 in scripts/build-images.sh: Declare and assign separately
- Fix SC2012 in scripts/generate-manifests.sh: Use find instead of ls
- Update Makefile to exclude vendored deskrun scripts from linting

Changes:
- scripts/build-images.sh: Split size variable declaration and assignment
- scripts/generate-manifests.sh: Replace ls with find for listing generated files
- Makefile: Lint only our scripts (scripts/*.sh and .github/scripts/tag-and-push-image.sh)
  excluding vendored scripts (nix-setup.sh, install-devbox.sh, install-docker.sh)

All shellcheck findings now resolved.

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: run tag-and-push-image.sh directly instead of through devbox run

The workflow was failing because devbox run interprets script names from
devbox.json, not bash commands. Since .github/scripts/tag-and-push-image.sh
has a proper shebang (#!/usr/bin/env bash) and execute permissions, it can
be invoked directly.

Changes:
- Remove "devbox run -- bash" prefix from tag/push steps
- Call .github/scripts/tag-and-push-image.sh directly with arguments
- Docker CLI is still available via install-docker.sh script

This fixes the error:
  bash: .github/scripts/tag-and-push-image.sh: No such file or directory
  Error: error running script "bash" in Devbox: exit status 127

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: add devenv workaround for cached runners

Add DEVENV_ROOT="" environment variable to disable devenv integration 
in nix, which causes issues on cached runners. Also clean stale devbox 
state before installing bob.

This follows the same workaround pattern used in instant-bosh:
https://github.com/rkoster/instant-bosh/blob/a9bc60ce521833606f16b5aa66c9f735ffc040e9/.github/workflows/build-and-publish.yml

Fixes: rkoster/rubionic-workspace#215

rubionic committed:
fix: add ./ prefix to tag-and-push-image.sh script calls

The workflow was failing with "No such file or directory" because bash
doesn't search the current directory for executables unless prefixed
with ./ or an absolute path.

Changes:
- Add ./ prefix to both database and runtime tag/push script calls
- This makes the path explicitly relative to current directory

Fixes: rkoster/rubionic-workspace#215
